src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -515,6 +502,54 @@ async def build_pex(
     )
 
 
+def _validate_metadata(
+    metadata: LockfileMetadata, request: PexRequest, python_setup: PythonSetup
+) -> None:
+    validation = metadata.is_valid_for(
+        request.requirements.lockfile_hex_digest,
+        request.interpreter_constraints,
+        python_setup.interpreter_universe,
+    )
+
+    if validation:
+        return
+
+    message_parts = [
+        f"Invalid lockfile for PEX request `{request.output_filename}`.",
+        "\n",
+    ]
+
+    if InvalidLockfileReason.INVALIDATION_DIGEST_MISMATCH in validation.failure_reasons:
+        message_parts.append(
+            "The requirements set for this PEX request are different to the requirements set when "
+            "the lockfile was generated. To fix this, you will need to regenerate the lockfile. "
+        )
+        message_parts.append(
+            f"(Expected requirements digest: {request.requirements.lockfile_hex_digest}, "
+            "actual: {metadata.requirements_invalidation_digest})"
+        )
+        message_parts.append("\n")
```

Comment:
This can use `.extend()`. Are you using `.append()` perhaps for the sake of formatting?